Azure/sonic-utilities
# Mpls cli

#### What I did

(PLEASE ONLY CHECK MY LATEST COMMIT)

Added bug fixes for MPLS CLI commands where mpls show and config commands would allow invalid interfaces to be passed into the command, without displaying an error. No longer require the "-d all" flag to show internal interfaces too, when a specific interface is asked to "show". Also added respective unit tests to check out the errors, and test the expected output. 

#### How I did it

Modified the config/main.py to add logic to detect invalid interfaces by comparing against PORTCHANNEL_INTERFACES, INTERFACE, and VLAN_INTERFACE. Similarly added logic in show/interfaces in __init__.py to detect invalid interfaces by comparing interfaces across all ASICs. 

#### How to verify it

Type the following commands into the CLI, and see the output:

"show interfaces mpls $VALIDPORTCHANNEL" -> should output a table format of whether that interface has MPLS enabled or not. Notice the "d-all" flag is not required here, an internal interface will also be displayed when called. 

"show interfaces mpls $INVALID_PORTCHANNEL" -> produces an error as invalid interface

"sudo config interface -n $QASIC# mpls add $INVALID_PORTCHANNEL" -> produces an error as invalid interface

#### Previous command output (if the output of a command-line utility has changed)

For configuring an invalid interface in an ASIC:
"""
admin@vlab-07:~$ sudo config interface -n asic2 mpls add PortChannel4011
"""

No error is produced, which it should produce. 

For showing an invalid interface in an ASIC:
"""
admin@vlab-07: show interfaces mpls PortChannel4001
Interface                 MPLS State
---------------       ------------
"""

An error should be produced, which it is not. 


#### New command output (if the output of a command-line utility has changed)

For configuring an invalid interface in an ASIC:
"""
admin@vlab-07:~$ sudo config interface -n asic2 mpls add PortChannel4011
Usage: config interface mpls add [OPTIONS] <interface_name>
Try "config interface mpls add -h" for help.

Error: interface PortChannel4011 doesn`t exist
"""

For showing an invalid interface in an ASIC:
"""
admin@vlab-07:~$ show interfaces mpls PortChannel3007
Error: Invalid interface. Interface not found.
"""

Valid show mpls interface:
"""
admin@vlab-07: show interfaces mpls PortChannel4001
Interface                 MPLS State
---------------       ------------
PortChannel4001   enable
"""

Notice the "-d all" flag is not required. Internal interfaces are also displayed when trying to show a specific interface.
